gamemodes/terrortown/entities/entities/ttt_base_placeable.lua

```diff
@@ -392,4 +392,43 @@ if SERVER then
 
         return true
     end
-end
+end -- SERVER
+
+if CLIENT then
+    ---
+    -- This is triggered, when you focus a marker of an entity and press 'Use'-Key
+    -- Override to handle remote use
+    -- @realm client
+    function ENT:RemoteUse() end
+
+    hook.Add("KeyRelease", "TTT2RemoteUsePlaceable", function(ply, key)
+        if
+            key ~= IN_USE
+            or not IsValid(ply)
+            or not ply:IsTerror()
+            or not IsFirstTimePredicted()
+        then
+            return
+        end
+
+        local focussedEnt = markerVision.GetFocussedEntity()
+        if not IsValid(focussedEnt) or not isfunction(focussedEnt.RemoteUse) then
+            return
+        end
+
+        -- see if we need to do some custom usekey overriding
+        local tr = util.TraceLine({
+            start = ply:GetShootPos(),
+            endpos = ply:GetShootPos() + ply:GetAimVector() * 110,
+            filter = ply,
+            mask = MASK_SHOT,
+        })
+
+        -- Check if we want to pickup a weapon or check a corpse
+        if tr.Hit and IsValid(tr.Entity) and (tr.Entity.CanUseKey or tr.Entity.player_ragdoll) then
+            return
+        end
+
+        focussedEnt:RemoteUse()
+    end)
+end -- CLIENT
```

Comment:
I know why you put it on the client, but should it really be on the client? Two things:

- `ENT:UseOveerride` is serverside
- there are quite a few role specific checks that can only be done serverside because the data is only valid on the server

My suggestion would complicate this whole system a bit, but:

- integrate this hook into the already existing system:
  https://github.com/TTT-2/TTT2/blob/2ae4f047ad44db09e214c87d5e9751a7c856a7ef/gamemodes/terrortown/gamemode/server/sv_player.lua#L523-L556
- call a function like `ENT:RemoteUseOverride` if it exists, that way it would also work for **all** entities, not just placeable entities
- populate the placeables UseOverride with a hook that can be overriden that contains a check if the player is able to open the UI, similar to this:
  https://github.com/TTT-2/TTT2/blob/bb659f30a2d3281ce6ccf1b0d5e1c8824c97363e/gamemodes/terrortown/entities/entities/ttt_base_placeable.lua#L245
- send a netmessage to the client that triggers a clientside RemoteUseOverride where the UI can be added.

But now that I think about this, this might be problematic in high ping scenarios. So see this as a starting point for a discussion. But I'm not sure that I like the current implementation that much.

Reply:
Or - if it is client side only like you implemented it currently - it should probably be a feature of markerVision, not the placeable entity. That way one could add UIs to all kind of things

Reply:
Well so here are the main points:
- Markervision is currently rendered and calculated clientside only.
- The ideal TTT2 GM:KeyRelease implementation for pickup and bodysearch is serverside only, but generally predicted and shared

So this was my idea to simply use that hook and just react to the usekey clientside, if a marker is focussed and do all the calls to the server from here.
It wouldnt make sense to combine this generally with the server as long as the server has no clue which markers you focus.
It would be much better to change our general use key system to a predicted one, that would also solve weapons not being properly picked up. But then we would also need to predict the marker system. We cant use eyetrace on the server, which is already quite bad with client vision.

Reply:
No, thats not what I meant to say. Just wanted to point out my problems. I agree, that this is not a good solution, but I wanted feedback, so thanks for that <3